bin/curriculum/delete_unit.rb

```diff
@@ -40,17 +40,22 @@ def parse_options
 
 require_relative '../../dashboard/config/environment'
 options.unit_names.each do |unit_nm|
+  puts "Processing unit [#{unit_nm}]..."
+
   unit_to_del = Unit.find_by(name: unit_nm)
   raise "Unit with name #{unit_nm} not found" if unit_to_del.nil?
-  raise "Only units in_development can be deleted as they wouldn't have usage and hence any DB references." unless unit_to_del.published_state == "in_development"
+  raise "Published units (state set as 'Stable') cannot be deleted as they could have active sections and progress." unless unit_to_del.published_state != "stable"
```

Comment:
the rules around published state are a bit non-intuitive and I'm concerned that this is no longer enforcing the rules that it looks like its enforcing. here's how published state works (a few examples):

1. courese csd-2023 has published state stable, course csd3-2023 has published state `nil` (for units in a course it is inherited from the course)
2. when you mark a unit as "hidden", we actually overload the published_state field to represent this. for example, right now csd-2024 has published_state preview, and csd6b-2024 has published_state "in_development" (because it was hidden).
3. standalone units like coursea-2023 use their own published state field to determine their published state

so my concern is that this makes it easier to delete a unit in a published course, while giving the impression that it is protecting against this case by checking the published state. 

so, the original code (perhaps accidentally?) requires the unit to either be hidden, or to be truly in_development.

if you did want to get the published state of a unit, agnostic of whether it is in a course, you could use `unit.get_published_state`.

I don't feel strongly about having especially strong protections here... but I would like whatever we include here to not be misleading.

any preferred options given this analysis?

Reply:
to clarify i don't feel strongly about strong protections because it seems more important whether a section is assigned to the unit than anything else.

Reply:
Thanks Dave for that input. I wasn't aware that published state for units can be inherited potentially from the overall course. Yes, the original was accidentally preventing a unit from getting deleted. 

I agree with your concern around misleading sense of protection because of the inheritance of published state between unit group. I see two potential options here, one to use the get_published_state to address this or not do any checks at all, but rather include a warning message asking the user of the script to confirm the script can be deleted.

I think doing a combination of both would be good, i.e., update the check to use get_published_state and also include a warning message, something like, "You are deleting a script that is not in_development (and this can extend in the future to include a set of safe states)  - please confirm with the curriculum team and involve CX to provide appropriate comms."  Thoughts?

Reply:
yes, sounds good to me, although there may be an extra step to NOT warn in the case where the unit is within a published unit group but has been marked as hidden.

Reply:
Updated to include the warning and use `get_published_state` - using the method also ensured the case where unit is within a published unit group, but marked hidden.

Including output from local run. microbit was in preview state, while the other two had state set in_development
```vijaya@VijiMiniDskTp:~/source/code-dot-org$ bin/curriculum/delete_unit.rb -u microbit-2024,focus-on-coding4b-2024,focus-on-creativity4b-2024 --dry-run
GetSecretValue: development/cdo/slack_bot_token
WARNING: MYSQL_OPT_RECONNECT is deprecated and will be removed in a future version.
Processing unit [microbit-2024]...
WARNING: MYSQL_OPT_RECONNECT is deprecated and will be removed in a future version.
Published state for Unit [microbit-2024] is not in_development, script might have usage. Please confirm with curriculum team and ensure comms has been sent before deletion. Continue? (Y/N)n
Processing unit [focus-on-coding4b-2024]...
No references to Unit [focus-on-coding4b-2024] found from sections. All validations for deletion passed.
Skipping deletion due to dry run mode.
Processing unit [focus-on-creativity4b-2024]...
No references to Unit [focus-on-creativity4b-2024] found from sections. All validations for deletion passed.
Skipping deletion due to dry run mode.